File: optimizers/Rprop.py
from .Optimizer import Optimizer, np


class Rprop(Optimizer):
    """
    Rprop — метод с индивидуальными адаптивными шагами по знакам градиента
    (resilient backpropagation).

    Первоисточник: Riedmiller M., Braun H. "A Direct Adaptive Method for Faster
    Backpropagation Learning: The RPROP Algorithm". IEEE International Conference
    on Neural Networks, 1993.
    https://doi.org/10.1109/ICNN.1993.298623
    Реализация следует алгоритму из документации PyTorch:
    https://docs.pytorch.org/docs/stable/generated/torch.optim.Rprop.html
    """

    param_descriptions = {
        "dec_factor": "множитель уменьшения длины шага при смене знака градиента (минимум проскочен)",
        "inc_factor": "множитель увеличения длины шага, пока знак градиента по координате не меняется",
        "step_min": "нижняя граница длины шага по каждой координате",
        "step_max": "верхняя граница длины шага по каждой координате",
    }

    last_gradient: np.ndarray
    step_size: np.ndarray

    def __init__(
        self,
        initial_x: np.ndarray,
        function,
        dec_factor: float = 0.5,
        inc_factor: float = 1.2,
        step_min: float = 1e-06,
        step_max: float = 50,
    ) -> None:
        params = dict(dec_factor=dec_factor, inc_factor=inc_factor, step_min=step_min, step_max=step_max)
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)

        mul = gradient * self.last_gradient
        factor = np.where(mul > 0, self.params["inc_factor"], np.where(mul < 0, self.params["dec_factor"], 1.0))
        self.step_size = np.clip(self.step_size * factor, self.params["step_min"], self.params["step_max"])
        gradient = np.where(mul < 0, 0.0, gradient)

        self.last_gradient = gradient
        next_x = self.x - np.sign(gradient) * self.step_size
        return self.move_next(next_x)

    def reset(self) -> None:
        super().reset()
        self.last_gradient = np.zeros([2])
        self.step_size = np.ones([2])

File: src/shared/lib/optimization-engine/optimizers/rprop.ts
import type { OptimizerDescriptor } from "./types";
import { zero2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient } from "@shared/lib/optimization-engine/functions";

export const rpropOptimizer: OptimizerDescriptor = {
  name: "Rprop",
  params: {
    decFactor: {
      default: 0.5,
      description: "множитель уменьшения длины шага при смене знака градиента (минимум проскочен)",
    },
    incFactor: {
      default: 1.2,
      description: "множитель увеличения длины шага, пока знак градиента по координате не меняется",
    },
    stepMin: { default: 1e-6, description: "нижняя граница длины шага по каждой координате" },
    stepMax: { default: 50, description: "верхняя граница длины шага по каждой координате" },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let lastGradient: Vec2 = zero2();
    let stepSize: Vec2 = [1, 1];
    return {
      get x() { return x; },
      params,
      next() {
        const grad = gradient(fn, x[0], x[1]);
        const step = (i: 0 | 1): { grad: number; size: number } => {
          const mul = grad[i] * lastGradient[i];
          const factor = mul > 0 ? params.incFactor : mul < 0 ? params.decFactor : 1.0;
          const size = Math.min(Math.max(stepSize[i] * factor, params.stepMin), params.stepMax);
          return { grad: mul < 0 ? 0.0 : grad[i], size };
        };
        const s0 = step(0);
        const s1 = step(1);
        stepSize = [s0.size, s1.size];
        lastGradient = [s0.grad, s1.grad];
        x = [
          x[0] - Math.sign(lastGradient[0]) * stepSize[0],
          x[1] - Math.sign(lastGradient[1]) * stepSize[1],
        ];
        return { x, value: fn(x[0], x[1]), internals: {
          "step.x": stepSize[0], "step.y": stepSize[1],
          "grad.x": lastGradient[0], "grad.y": lastGradient[1],
        } };
      },
      reset() {
        x = initialX;
        lastGradient = zero2();
        stepSize = [1, 1];
      },
    };
  },
};
